Permission Refiner

ivawzh opened the pull request:
Rough documentation done.
Unit tests done.
Supports array of ref. e.g. `locations:ref` & `locations: [refs]`
Haven't implement it to integration test to run the full tests yet.
fixes #6 

---------------------

TODO:
- [ ] support population
- [ ] Review Readme

---------------------
###It won't support:

Module's `READ` check on CREATE and UPDATE. For example, if the Audit Entry module is set to 'CREATE, UPDATE, DELETE' without READ. After a user created the Entry, the returned object will show him all the fields (if we don't define any separate field permissions) of the object. 
 
The reason why it can't support module check is because there is no information about module's permission on `permissionOptions`, i.e. `permissionOptions: {filters:{}, fields:[]}`.

willmcclellan commented:
I've only had a quick look through this, but two things that are confusing me:

* Why is this a mongoose plugin? Did we not discuss that this could work on a plain old JavaScript object and there for be agnostic to the ORM you are using? What are the benefits of this being a JS plugin?
* Why have you written e2e tests with the mock express app again? I'm sure we spoke about this already, but by doing this we are not increasing our coverage whatsoever. You are basically testing the fake express app you have setup which is not part of the module that people will use

ivawzh commented:
Right, I have removed the integration test. All tests are unit tests now. I have also changed the refiner function into an individual function. It doesn't rely on mongoose's namespace conventions anymore. =]

willmcclellan commented on README.md:
```diff
@@ -172,9 +172,84 @@ module.exports = (function byApplication(req, res, next) {
+###Use Permissions Refiner in your controller
+
+Permission Refiner will hide the fields which user doesn't have permission to `READ`;
+hide the complete content and return empty String '' if the user doesn't have `READ` on this document or its dependency.
```
Is the behaviour to set the field to an empty string, or remove it from the object?

ivawzh replied:
now it will return a null, when user doesn't have permission on read but have permission on update/create

willmcclellan commented on README.md:
```diff
@@ -172,9 +172,84 @@ module.exports = (function byApplication(req, res, next) {
+	};
+};
+
+module.exports = update;
```
I think we're ok to right our docs in ES6 as well

ivawzh replied:
Cool

willmcclellan commented on lib/post-save-refiner/index.js:
```diff
@@ -0,0 +1,116 @@
+import _ from 'lodash'
+import {isEmpty} from 'lodash'
```
2 imports of lodash here, syntax should be:

`import { isEmpty, isObject, ... } from 'lodash'

willmcclellan replied:
Also should have a space between your braces and the variable

willmcclellan commented on lib/post-save-refiner/index.js:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  // for filters
+  const filters = permissionOptions.filters
```
`const { filters, fields } = permissionOptions`

willmcclellan commented on lib/post-save-refiner/index.js:
```diff
@@ -0,0 +1,116 @@
+  const thisModule = modelName.toLowerCase()
+  if (!_.isEmpty(filters) && _.isObject(filters)) {
+    if (matchFilters(filters, doc, thisModule, options)) {
+      return '' // return empty string if no read permission on this doc/ its dependency
```
Yeh I'm not sure returning an empty string is the correct behaviour. Do you think deleting the property is more appropriate?

willmcclellan commented on lib/post-save-refiner/index.js:
```diff
@@ -0,0 +1,116 @@
+  // for filters
+  const filters = permissionOptions.filters
+  const thisModule = modelName.toLowerCase()
+  if (!_.isEmpty(filters) && _.isObject(filters)) {
```
What's the significance of this conditional... why are we checking for an object?

ivawzh replied:
> _.isEmpty([1, 2, 3]);
> // → false

So I was making sure it is not an array.

willmcclellan commented on lib/post-save-refiner/index.js:
```diff
@@ -0,0 +1,116 @@
+
+  // for fields
+  // clone the doc to avoid pollution
+  const newDoc = _.cloneDeep(doc)
```
I'm wondering if there's any advantage to this? Immutability can be a nice pattern, but it doesn't guard against anything here from what I can see?

willmcclellan commented on lib/post-save-refiner/index.js:
```diff
@@ -0,0 +1,116 @@
+    // fields originally looks like this: ['-name', '-number', '-object']
+    const field = minorsField.replace(/^-/, '') // strip `-` sign
+    if (_.has(newDoc, field)) {
+      delete newDoc[field]
```
How does this work with a nested field reference?

```
const fields = ['-address.street']
const obj = {
  address: {
    street: 'Easey Street'
  }
}

delete obj['address.street']

willmcclellan commented on lib/post-save-refiner/index.js:
```diff
@@ -0,0 +1,116 @@
+
+/**
+ * Refine the returned document from a `save` action(i.e. `update` and `create`) by its permission rules.
+ * @param {object} doc - The document that is going to be refined.
```
I think `doc` should be `obj` now that this isn't a mongoose plugin?

willmcclellan commented on lib/post-save-refiner/test/refine-test.js:
```diff
@@ -0,0 +1,326 @@
+import should from 'should'
+import permissionsRefine from '../index'
+import mongoose from 'mongoose'
+const ObjectId = mongoose.Types.ObjectId
```
I think we can remove any reference to mongoose/ObjectId from these tests. Just to make sure there is no confusion around any reliance on mongoose/mongo. Your `_id`s can be any unique string

willmcclellan commented on lib/post-save-refiner/test/refine-test.js:
```diff
@@ -0,0 +1,326 @@
+    }
+  }
+
+  it('should not change the doc if filters:blank, fields:*', done=> {
```
My OCD is kicking in, can we add a space before any fat arrows

willmcclellan commented on lib/post-save-refiner/test/refine-test.js:
```diff
@@ -0,0 +1,326 @@
+      fields: ['-array', '-number']
+    }
+    const output = permissionsRefine(doc, permissionsOptions, model, options)
+    output.should.eql('')
```
This is an interesting one... What do we return when permission fails for the object. Empty string, blank object, error?

willmcclellan commented:
How about we go with the name, `sanitize` for the function. I think that represents what it's job is quite well, sanitizing an object based  on permissions you pass in...

So the function can live in: 'lib/sanitize'

Usage would be like...

```
import { sanitize } from 'hall-pass'

const sanitizedObj = sanitize(obj, permissions, module, options) 
```

willmcclellan commented on lib/post-save-refiner/index.js:
```diff
@@ -0,0 +1,115 @@
+ * Refine the returned document from a `save` action(i.e. `update` and `create`) by its permission rules.
+ * @param {object} doc - The document that is going to be refined.
+ * @param {object} permissionOptions - A object, which is generated by function `ask`, contains permission information
+ * @param {string} modelName - The model name of the target document
```
should this be `moduleName` like what is referenced in the options?

willmcclellan commented on lib/sanitize/index.js:
```diff
@@ -0,0 +1,145 @@
+  return any(arrayOne, element => {
+    return includes(arrayTwo, element)
+  })
+}
```
https://lodash.com/docs#intersection

willmcclellan commented:
:+1:
